src/main/java/uk/gov/pay/api/json/RequestJsonParser.java

```diff
@@ -214,10 +217,22 @@ private static String validateAndGetAgreementReference(JsonNode agreementRequest
     }
 
     private static String validateAndGetReference(JsonNode request, RequestError.Code validationError, RequestError.Code missingFieldError) {
-        return validateAndGetString(
+        String reference = validateAndGetString(
                 request.get(REFERENCE_FIELD_NAME),
                 aRequestError(REFERENCE_FIELD_NAME, validationError, "Must be a valid string format"),
                 aRequestError(REFERENCE_FIELD_NAME, missingFieldError));
+        validateNoIllegalCharacters(reference, REFERENCE_FIELD_NAME, validationError);
+        return reference;
+    }
+
+    private static void validateNoIllegalCharacters(String fieldValue, String fieldName, RequestError.Code validationError) {
+        if (fieldValue != null) {
+            for (char c : fieldValue.toCharArray()) {
+                if (NAXSI_NOT_ALLOWED_CHARACTERS.contains(c)) {
+                    throw new BadRequestException(aRequestError(fieldName, validationError, "Must be a valid string format"));
+                }
+            }
```

Comment:
No need to convert it to a character array: `for (int i = 0; i < fieldValue.length(); i++)` (with `fieldValue.charAt(i)` inside `contains(…)` on the next line) would do it. (It would be nice to do this with a stream but unfortunately the `chars()` method on `String` returns an `IntStream` of code points rather than a stream of characters.)

Reply:
Amended, thanks Alex! :)